shared/js/background/components/abn-experiments.js

```diff
@@ -0,0 +1,221 @@
+     * }} opts
+     */
+    constructor({ abnMetrics }) {
+        browser.webRequest.onCompleted.addListener(
+            async (details) => {
+                await abnMetrics.remoteConfig.ready;
+                const url = new URL(details.url);
+                if (url.pathname.startsWith('/t/epbf_')) {
+                    // broken site report
+                    abnMetrics.onMetricTriggered('brokenSiteReport');
+                }
+            },
+            {
+                urls: ['https://improving.duckduckgo.com/t/*'],
+                tabId: -1,
+            },
+        );
+    }
+}
+
+/**
+ * A/B/N testing framework: metrics.
+ *
+ * This component handles metrics needed for A/B testing. It handles the rate-limiting required
+ * to define and send metrics as per the documentation in Asana (https://app.asana.com/0/1208889145294658/1208747415972722/f)
+ *
+ * Usage:
+ *  - The main entry point is `markExperimentEnrolled`. This should be used to move an assigned experiment to
+ * 'enrolled' state when the user first sees the treatment behavior. You can call this multiple times, as
+ * subsequent calls will be a no-op if the user is already enrolled.
+ *  - When enrolling, you can pass a list of 'metrics' that should be sent for the experiment. Each metric defines an
+ * event, value count and conversion window. e.g. metric: 'search', value: 4, conversionWindow 1-7: triggered
+ * once 4 searches have been done within 1-7 days (inclusive) of enrollment.
+ *  - `generateRetentionMetrics` can be used to create a default list of retention metrics (search and app_use).
+ *
+ * Example:
+ *  - Check if user has been assigned a cohort for the subfeature: `remoteConfig.getCohortName(feature, subFeature) !== null`
+ *  - Check if user is in the 'treatment' group: `remoteConfig.isSubFeatureEnabled(feature, subFeature, 'treatment')`
+ *  - Enroll the user in the experiment (with default retention metrics.): `abnMetrics.markExperimentEnrolled(feature, subFeature)`
+ */
+export default class AbnExperimentMetrics {
+    /**
+     *
+     * @param {{
+     *  remoteConfig: import('./remote-config').default
+     * }} opts
+     */
+    constructor({ remoteConfig }) {
+        this.remoteConfig = remoteConfig;
+    }
+
+    /**
+     * Mark that the user should be enrolled in the experiment for the given feature and sub feature.
+     *
+     * This will send the enrollment
+     * @param {string} featureName
+     * @param {string} subFeatureName
+     * @param {ExperimentMetric[]} [metrics]
+     */
+    markExperimentEnrolled(featureName, subFeatureName, metrics) {
+        const cohort = this.remoteConfig.getCohort(featureName, subFeatureName);
+        if (cohort && !cohort.enrolledAt) {
+            cohort.enrolledAt = Date.now();
+            cohort.metrics = (metrics || generateRetentionMetrics()).map((m) => ({ ...m, counter: 0, sent: false }));
+            sendPixelRequest(`experiment_enroll_${subFeatureName}_${cohort.name}`, {
```

Comment:
Are we sure that `subFeatureName` and `cohort.name` will never have special characters that could screw up the pixel request's URL? If not (or just to be safe), we could encode the pixel name in `sendPixelRequest()`?

Reply:
`subFeatureName` must conform to `/^[a-zA-Z0-9]+$/`, which is enforced via unit-tests in the privacy config: https://github.com/duckduckgo/privacy-configuration/blob/main/tests/config-tests.js#L68-L75

`cohort.name` doesn't have the same validation, so could be problematic. I'll check with native teams to see if they are doing some sanitization on the client, or we should just add config validation to protect it.

Reply:
Thanks, glad you're checking that with them. Perhaps it's also worth adding a `encodeURIComponent(pixelName)` call in the shared `sendPixel()` function to help protect other folks writing pixels in the future too?

Reply:
Added config validation in https://github.com/duckduckgo/privacy-configuration/pull/2688.

I don't think adding `encodeURIComponent` in sendPixel will be of much use:
- It doesn't protect against `_`, which gets converted to a namespace separator on the backend.
- We generally shouldn't be sending variables we don't control in the pixel name, so this would only be catching an suppressing coding mistakes, which could end up hiding bugs. I'd rather fail fast here.